### DISCUSSION TITLE
Explicit structure layout and constructor

### DISCUSSION THREAD

**Discussion by levicki** [dotnet/csharplang#7261] [Q&A] [answered]
Please consider this C# code:
```
public static class MSSIGN32
{
	// SIGNER_SUBJECT_INFO.dwSubjectChoice
	private const UInt32	SIGNER_SUBJECT_FILE = 1;
	private const UInt32	SIGNER_SUBJECT_BLOB = 2;

	[StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
	private struct SIGNER_FILE_INFO
	{
		public UInt32		cbSize;
		public string		Filename;
		public IntPtr		Handle;

		public SIGNER_FILE_INFO(string _Filename)
		{
			cbSize = (UInt32)Marshal.SizeOf(typeof(SIGNER_FILE_INFO));
			Filename = _Filename;
			Handle = IntPtr.Zero;
		}
	}

	[StructLayout(LayoutKind.Explicit, CharSet = CharSet.Unicode)]
	private struct SIGNER_SUBJECT_INFO
	{
		[FieldOffset(0)]
		public UInt32		cbSize;
		[FieldOffset(4)]
		public IntPtr		Index;
		[FieldOffset(12)]
		public UInt32		SubjectChoice;
		[FieldOffset(16)]
		public IntPtr		SignerFileInfo;
		[FieldOffset(16)]
		public IntPtr		SignerBlobInfo;

		public SIGNER_SUBJECT_INFO(UInt32 _Index, SIGNER_FILE_INFO _SignerFileInfo)
		{
			cbSize = (UInt32)Marshal.SizeOf<SIGNER_SUBJECT_INFO>();
			Index = Marshal.AllocHGlobal(Marshal.SizeOf<UInt32>());
			Marshal.WriteInt32(Index, (Int32)_Index);
			SubjectChoice = SIGNER_SUBJECT_FILE;
			SignerFileInfo = Marshal.AllocHGlobal(Marshal.SizeOf<SIGNER_FILE_INFO>());
			Marshal.StructureToPtr(_SignerFileInfo, SignerFileInfo, false);
		}
	}
}
```
It results in the following error message:
```
error CS0171: Field 'MSSIGN32.SIGNER_SUBJECT_INFO.SignerBlobInfo' must be fully assigned before control is returned to the caller. Consider updating to language version '11.0' to auto-default the field.
```
Am I wrong in expecting one of the following things to be true:

1. Either the non-default constructor is not allowed in a struct marked with explicit layout
2. Or this error is a compiler bug since it should know from the `FieldOffset` that the `SignerBlobInfo` has been initialized indirectly

Please advise.

**Comment by 333fred** [comment]
The compiler doesn't know anything about FieldOffset or take it into account at all. To silence this error/warning (depends on which C# version you're in), I would suggest `this = default;` at the start of the ctor.

**Reply by FaustVX** [reply]
Or `: this()` should also work on older version ?

**Reply by 333fred** [reply]
That would also work, yes.

**Reply by levicki** [reply]
@333fred 

> To silence this error/warning (depends on which C# version you're in)...

C# version is 7.3 (.Net Framework 4.8) and it is an error, not a warning (if it were a warning I wouldn't bother asking about it).

> ... I would suggest this = default; at the start of the ctor.

Two questions:

1. Wouldn't that be inefficient (initializing everything to 0/null/IntPtr.Zero or whatever default) only to get overwritten right away? I am unaware of the state of C# compiler code optimization when it comes to such things.
2. If both values are initialized which one will be taken by the marshaler? First listed? How does that work when structure is not marked as sequential?

> The compiler doesn't know anything about FieldOffset or take it into account at all.

Then perhaps it should if `FieldOffset` is the accepted way of specifying overlapping fields in a structure?

Marshaling structs with unions is part of P/Invoke and I don't see it going away anytime soon. I think that compiler should know that those fields with same offset are aliases and not throw a tantrum.

**Reply by huoyaoyuan** [reply, marked as answer]
> 1. Wouldn't that be inefficient (initializing everything to 0/null/IntPtr.Zero or whatever default) only to get overwritten right away?

You can use `Unsafe.SkipInit(out this)` to skip the check.

> 2. If both values are initialized which one will be taken by the marshaler?

FieldOffset is often handled by the runtime. Only one value will be stored.

BTW your P/Invoke code is wrong, because SIGNER_FILE_INFO and SIGNER_BLOB_INFO should be pointers instead of nested structs. I'd suggest using [CsWin32](https://www.nuget.org/packages/Microsoft.Windows.CsWin32) to auto generate P/Invoke code for correctness.

**Reply by levicki** [reply]
@huoyaoyuan 
> You can use `Unsafe.SkipInit(out this)` to skip the check.

Thanks for pointing me in the right direction. [This documentation](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.unsafe.skipinit?view=net-8.0) has an example and it was just what I needed to see to understand how this works.

> BTW your P/Invoke code is wrong...

Yes, I am aware that it should have been `IntPtr`, not the struct itself &mdash; I was in a hurry when I posted the question. I have changed the code example to reflect reality.

> I'd suggest using [CsWin32](https://www.nuget.org/packages/Microsoft.Windows.CsWin32) to auto generate P/Invoke code for correctness.

Thanks for the suggestion, but I can marshal just fine on my own and usually without having to resort to `unsafe` blocks and certainly without requiring C# version 9 (I still have to maintain code using .Net Framework 4.x).

On a side note, for any automatic marshaling code generator to work reliably, the API documentation from which the method signatures are pulled must be correct, which for say [SignerTimeStampEx2](https://learn.microsoft.com/en-us/windows/win32/seccrypto/signertimestampex2) as well as [SignerTimeStampEx3](https://learn.microsoft.com/en-us/windows/win32/seccrypto/signertimestampex3) is not true (`ALG_ID dwAlgId` should be `PCSTR pszAlgorithmOid` for the former, and `PCWSTR pszAlgorithmOid` should be `PCSTR pszAlgorithmOid` for the latter). I doubt automatic marshaler would have figured that one out unless it has some internal source of correct API signatures.
